### ISSUE TITLE
Feature Request: Play video on Phone from Kodi Library

### ISSUE THREAD

**Issue by kirbyzhou** [xbmc/Official-Kodi-Remote-iOS#702]
Now, Kore for Android can play video on Phone from Kodi Library.
Can [Official-Kodi-Remote-iOS](https://github.com/xbmc/Official-Kodi-Remote-iOS) also add this feature?

**Comment by wutschel** [comment]
There was an unfinished attempt to play file via VLC, the code got removed. Adding this feature is not on my agenda yet, but I understand this can be useful. Any help to add this is appreciated.

**Comment by wutschel** [comment]
Just did a quick hack, and this already plays my Flac audio files, but ir refuses to play the mpg and mkv files I have. I guess `AVPlayer` does have limited support for A/V formats.

https://github.com/wutschel/Official-Kodi-Remote-iOS/tree/local_playback

**Comment by kambala-decapitator** [comment]
> I guess `AVPlayer` does have limited support for A/V formats.

yes, you're right.

I think it'd be better to add a Share button sending video URL or so that can be transferred to an external video player app capable of handling such format.

**Comment by wutschel** [comment]
Why not. Added there long press action "Share" which for now only allows to copy-to-clipboard.

https://github.com/wutschel/Official-Kodi-Remote-iOS/tree/share_url

**Comment by wutschel** [comment]
Thanks for your comments. I just started to play around with this a bit and will improve this further. Finally I need to test on my phone as on simulator I don't have any other apps running.

**Comment by wutschel** [comment]
Updated the implementation, mainly addressing the origin and not disabling mail/message/airdrop. Can you help with how to update icon? I did not find reference obj c code around `activityViewControllerLinkMetadata` and `UIActivityItemProvider`.

**Comment by kambala-decapitator** [comment]
here's a URL wrapper from one of my projects, hope this would be sufficient to get the idea:

```objc
// .h
@interface SharingActivityItemSource : NSObject <UIActivityItemSource>
- (instancetype)initWithUrlString:(NSString *)urlString;
@end

// .m
@import LinkPresentation;

@interface SharingActivityItemSource ()
@property (nonatomic, copy) NSURL *url;
@end

@implementation SharingActivityItemSource

- (instancetype)initWithUrlString:(NSString *)urlString {
    if (self = [super init]) {
        self.url = [NSURL URLWithString:urlString];
    }
    return self;
}

- (id)activityViewControllerPlaceholderItem:(UIActivityViewController *)activityViewController {
    return self.url;
}

- (id)activityViewController:(UIActivityViewController *)activityViewController itemForActivityType:(NSString *)activityType {
    return self.url;
}

- (LPLinkMetadata *)activityViewControllerLinkMetadata:(UIActivityViewController *)activityViewController API_AVAILABLE(ios(13.0)) {
    __auto_type meta = [LPLinkMetadata new];
    meta.originalURL = self.url;
    meta.URL = meta.originalURL;
    meta.title = @"Artikel";
    meta.imageProvider = [[NSItemProvider alloc] initWithContentsOfURL:NSBundle.mainBundle.appIconURL];
    return meta;
}

@end
```

and then you pass an instance of this class instead of plain URL to `activityItems`.

iirc there're more fields in LPLinkMetadata, please check its doc/header.

**Comment by wutschel** [comment]
Thanks, tried to add this to my best knowledge (see https://github.com/wutschel/Official-Kodi-Remote-iOS/tree/share_url), but this crashes with "*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[SharingActivityItemSource copyWithZone:]: unrecognized selector sent to instance 0x600001b00620'". If I implement it crashes with the next unrecognized selector. Did I miss something?

**Comment by kambala-decapitator** [comment]
you made a mistake: `activityItems` must be an array :)

**Comment by wutschel** [comment]
That was it :)

<a href="https://abload.de/image.php?img=bildschirmfoto2022-103ledk.png"><img src="https://abload.de/img/bildschirmfoto2022-103ledk.png" /></a>

**Comment by kambala-decapitator** [comment]
and now with LPLinkMetadata you can control title (`Down and out` instead of IP), possibly display album cover image etc.

**Comment by wutschel** [comment]
> and now with LPLinkMetadata you can control title (`Down and out` instead of IP), possibly display album cover image etc.

Yep, next step. But this might be a viable overall solution. Still need to test how this behaves when sharing the URL with other apps. Do you have by chance the possibility to test?

**Comment by kambala-decapitator** [comment]
yes, will try on iPad, have a few video players there

**Comment by wutschel** [comment]
Updated, but now sometimes not showing up on iPod. Will be related to the ctrl which shows the popup. Need to dig into this later.

<a href="https://abload.de/image.php?img=bildschirmfoto2022-10dhdbc.png"><img src="https://abload.de/img/bildschirmfoto2022-10dhdbc.png" /></a>

**Comment by kambala-decapitator** [comment]
tested on iPad: seems to work, opening the URL in VLC works as well. Unfortunately, plain URL can't be sent to video players out of the box, probably unless they implement a share extension.

`dummyView` - please don't do that. Simple download can be done with `NSURLSession`.

But now I see that thumbnails are usually unavailable, I'd avoid delays in showing the sharing view as it's a rather bad UX (at first I thought that sharing simply doesn't work and tried it a couple of times). Can be shown only if already available in local cache (if there's such thing in the Remote).

**Comment by wutschel** [comment]
dummyView was just to play with this -- I still see this as kind of prototyping. Yes, there is a thumbnail cache. If I use this, the  behaviour is different of course -- if cached, the thumbnail is shown; if not, the app icon is shown.

**Comment by wutschel** [comment]
Just updated and refined the prototype implementation in https://github.com/wutschel/Official-Kodi-Remote-iOS/tree/share_url

Should this feature (sharing download URL) be added, or is it too awkward to use (usually you need to copy-paste the URL into another player on your device)?

**Comment by kambala-decapitator** [comment]
> Should this feature (sharing download URL) be added, or is it too awkward to use (usually you need to copy-paste the URL into another player on your device)?

IMO wouldn't hurt :)
